Test/review requests

alexanderleegs opened the pull request:
This PR introduces tests for review requests. Resolves IS-261.

This PR also adds additional cypress utils and direct api calls to assist with testing. This includes a new collaborator test user `E2E_EMAIL_COLLAB_2`, which requires the domain `@e2etest.com` to be whitelisted on staging as well with a defined expiry.

Note that this test suite has code for inspecting the review request summary, but this test suite is currently skipped - the reason for this is it appears that the underlying github operations are flaky - attempting to add a file `blah.md` soon after deleting the same file `blah.md` instead returns a conflict error from github (saying that the file has been recently changed) - this occurs for several github operations. However, in order to maintain consistency in our test suites, we need to be able to clean up old test files created for this purpose, as well as set up similar test files for use - the end result is that it is very difficult to create a consistent test suite for this case, and as such I have opted to skip it for now, and file an issue to look into this in future.

### Before merging

- [ ] add `@e2etest.com` to our domain whitelist staging db

seaerchin commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
       // Arrange
+
       // Act
+      cy.visit(`/sites/${E2E_EMAIL_TEST_SITE.repo}/editPage/example-page.md`)
```
could we additionally check that pointer events are disabled? (lmk if this is hard)

alexanderleegs replied:
Added in e47b68e4

seaerchin commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+      openReviewRequestStateDropdown("Approved")
+      cy.contains("In review").should("exist")
+      cy.contains("In review").click()
+      // Wait for dropdown menu to disappear
+      cy.wait(1000)
```
could we do this by means of an assertion? ie, we get a selector that matches the dropdown menu and alias it (while it exists) then later on we do a `cy.get(dropdown_selector).should('not.exist')`

alexanderleegs replied:
Sure, fixed in https://github.com/isomerpages/isomercms-frontend/pull/1379/commits/8a4f085a55e109fa8c9081b30f8561d72b1def44!

seaerchin commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+      getViewReviewRequestButton().click()
+      cy.wait(`@${RR_INTERCEPTOR}`)
+      openReviewRequestStateDropdown("Approved")
+      cy.contains("In review").should("exist")
+      cy.contains("In review").click()
```
supernit: consider chaining

alexanderleegs replied:
Fixed in https://github.com/isomerpages/isomercms-frontend/pull/1379/commits/8a4f085a55e109fa8c9081b30f8561d72b1def44

seaerchin commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+      const RR_INTERCEPTOR = "getReviewRequest"
+      cy.intercept("GET", "**/summary").as(RR_INTERCEPTOR)
+      const RR_CANCEL_INTERCEPTOR = "getReviewRequest"
+      const regex = /\/review\/\d+\//
```
sorry, could i get a brief explanation on this?

alexanderleegs replied:
Have added a comment in 74c94e5f!

seaerchin commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+      const regex = /\/review\/\d+\//
+      cy.intercept("DELETE", regex).as(RR_INTERCEPTOR)
+      cy.setEmailSessionDefaults("Email admin")
+      cy.actAsEmailUser(E2E_EMAIL_ADMIN.email, "Email admin")
+      editUnlinkedPage(
+        "faq.md",
+        "some requestor close review request content",
+        E2E_EMAIL_TEST_SITE.repo
+      )
+      createReviewRequest(
+        MOCK_REVIEW_TITLE,
+        [E2E_EMAIL_ADMIN_2.email],
+        MOCK_REVIEW_DESCRIPTION
+      )
```
this pattern seems to be repeated q often throughout this suite - wdyt of extracting it as a util function in this file?

alexanderleegs replied:
Yeah that's a good point, shifted in e47b68e4!

seaerchin commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+      getViewReviewRequestButton().click()
+      cy.wait(`@${RR_INTERCEPTOR}`)
+      openReviewRequestStateDropdown("In review")
+      cy.contains("Cancel request").should("exist")
+      cy.contains("Cancel request").click()
```
supernit: consider chaining

alexanderleegs replied:
Fixed in https://github.com/isomerpages/isomercms-frontend/pull/1379/commits/8a4f085a55e109fa8c9081b30f8561d72b1def44

seaerchin commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+        cy.visit(`/sites/e2e-email-test-repo/review/${id}`)
+        // Assert
+        cy.contains(
+          "Please ensure that you have selected a valid review request."
```
could we check that it additionally redirects them?

alexanderleegs replied:
Sure, 734f4d22

seaerchin commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+        // Act
+        cy.visit(`/sites/e2e-email-test-repo/review/${id}`)
+        // Assert
+        cy.contains(
```
same as above - wdyt about checking for redirection?

alexanderleegs replied:
734f4d22

alexanderleegs commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -41,8 +63,26 @@ const REVIEW_MODAL_SUBTITLE =
+    cy.setupDefaultInterceptors()
     cy.setEmailSessionDefaults("Email admin")
     closeReviewRequests()
+    editUnlinkedPage(
```
74c94e5f

alexanderleegs commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -53,9 +93,29 @@ describe("Review Requests", () => {
 
+    it("should not be able to create a review request when there are no changes", () => {
+      // Arrange
+      // TODO: also create a new rr and merge it beforehand to make sure that there are no changes
```
Whoops it isn't because it's moved to directly after the `before`, which ensures that the state is consistent - removed in 74c94e5f

alexanderleegs commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -53,9 +93,29 @@ describe("Review Requests", () => {
+      visitE2eEmailTestRepo()
+      removeOtherCollaborators()
+      removeOtherCollaborators()
       removeOtherCollaborators()
```
This method is actually incorrectly named, it only removes a single collaborator, and I couldn't figure out how to fix it to remove all. We want to leave only the admin as the only collaborator to start the test case, and we have 4 users total, so this removes all others. I didn't want to modify it because other E2E prs might be using it - but after a quick check no one else seems to be using it, so it's been changed in 4e363ff0

alexanderleegs commented on cypress/utils/collaborators.ts:
```diff
@@ -21,15 +21,14 @@ export const getCollaboratorsModal = (): Cypress.Chainable<
 }
 
 export const removeOtherCollaborators = (): void => {
```
makes sense, 4e363ff0

alexanderleegs commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -53,9 +93,29 @@ describe("Review Requests", () => {
+      cy.setEmailSessionDefaults("Email admin")
+      visitE2eEmailTestRepo()
+      removeOtherCollaborators()
+      removeOtherCollaborators()
```
We want to leave only the admin as the only collaborator to start the test case, and we have 4 users total, so this removes all others. The method has been renamed in 4e363ff0 to make this clearer

alexanderleegs commented on cypress/fixtures/constants.ts:
```diff
@@ -10,10 +10,18 @@ export const E2E_EMAIL_ADMIN = {
 } as const
 
+export const E2E_EMAIL_ADMIN_2 = {
+  email: "twodmin@e2e.gov.sg",
```
Haha i'll leave the existing name as is then!

I was facing issues with the FE flow because adding a collaborator has a secondary modal to go through - I opted to make a new user where the emails are split more explicitly instead!

alexanderleegs commented on cypress/fixtures/constants.ts:
```diff
@@ -10,10 +10,18 @@ export const E2E_EMAIL_ADMIN = {
   email: "collab@e2e.gov.sg",
 } as const
 
+export const E2E_EMAIL_COLLAB_2 = {
```
2a1b0244

alexanderleegs commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
       // select reviewer and click
-      selectReviewer(E2E_EMAIL_COLLAB.email)
+      selectReviewer(E2E_EMAIL_ADMIN_2.email)
       removeReviewers()
       getSubmitReviewButton().should("be.disabled")
       // add back reviewer
       getAddReviewerDropdown().click()
-      selectReviewer(E2E_EMAIL_COLLAB.email)
+      selectReviewer(E2E_EMAIL_ADMIN_2.email)
```
This was to ensure that remove flow is working fine!

alexanderleegs commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+      visitE2eEmailTestRepo()
+      closeReviewRequests()
     })
-    it("should have the pending review alert on the workspace", () => {
```
Yup, it was just renamed to `should have the review request approved alert stating that editing is disabled when on the workspace` to make the test clearer

alexanderleegs commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
-          ).then((id) => approveReviewRequest(id))
-        }
+      addAdmin(E2E_EMAIL_ADMIN_2.email)
+      closeReviewRequests()
```
This is in case of permissions issues - one possible area of inconsistency is that a test suite could be run in which the RR was created by a different user (e.g. ADMIN_2), then failed with the RR still open - when rerunning the test, the close would fail because the test user is not the creator of the RR

alexanderleegs commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+    it("should have the review request approved alert stating that editing is disabled when on the workspace", () => {
       // Arrange
+      const RR_INTERCEPTOR = "getReviewRequest"
+      cy.intercept("GET", "**/summary").as(RR_INTERCEPTOR)
```
Refactored in e47b68e4!

alexanderleegs commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+
       // Assert
+      // Toast
+      cy.contains("There is currently an approved review request!")
```
Fixed in e47b68e4

alexanderleegs commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+          }
+        }
+      )
+      // Wait for github to update and clear local storage - attempting to create a recently deleted file will return a 409 error on github's end
+      cy.wait(60000)
```
Ah this was just from internal testing, I tried a few values but was getting 409s from github - it actually doesn't even work consistently with this time interval, which is why i skipped this test suite

alexanderleegs commented on cypress/e2e/reviewRequests.spec.ts:
```diff
@@ -108,127 +157,510 @@ describe("Review Requests", () => {
+      openReviewRequestStateDropdown("In review")
+
       // Assert
+      cy.contains("Approved").should("exist")
```
This test case is actually to make sure collaborator-opened RRs can still be approved as per normal! I've changed the description, as well as added the case for collab being unable to approve RR in https://github.com/isomerpages/isomercms-frontend/pull/1379/commits/21c3af6a45ab8943d977cdd0698a5194ee5ed830

seaerchin reviewed (approved):
there are no blockers for functionality in this PR. however, there are some clean-ups that we could do, so that bad code practices don't persist in our codebase.

1. for the repeated `removeFirstCollaborator`, could we shift this into a util function (in the same file) and give it a coherent name + doc (i think the original `removeOtherCollaborators` work, but we need a comment explaining that there are exactly 3 others). 
2. For the commands, could we shift into `utils/reviewRequests` rather than having it on `cy`? (see [here](https://www.notion.so/opengov/E2E-email-tests-e8d088d60fb3492686451999fe75d76b?pvs=4#7a55593953544576811bc34d18ba37e3) for reasoning). 

i'll approve this but please do have them in a follow-up PR!

alexanderleegs commented:
> there are no blockers for functionality in this PR. however, there are some clean-ups that we could do, so that bad code practices don't persist in our codebase.
> 
> 1. for the repeated `removeFirstCollaborator`, could we shift this into a util function (in the same file) and give it a coherent name + doc (i think the original `removeOtherCollaborators` work, but we need a comment explaining that there are exactly 3 others).

Added in https://github.com/isomerpages/isomercms-frontend/pull/1379/commits/1ff2eeaaa3716037931c82740756d0b60d379c4e

> 2. For the commands, could we shift into `utils/reviewRequests` rather than having it on `cy`? (see [here](https://www.notion.so/opengov/E2E-email-tests-e8d088d60fb3492686451999fe75d76b?pvs=4#7a55593953544576811bc34d18ba37e3) for reasoning).

fixed in https://github.com/isomerpages/isomercms-frontend/pull/1379/commits/93462810c21a534acdb9a5dc89d7d82b88addd60


`closeReviewRequests()` works regardless of review request status - it retrieves the list of existing RRs first and closes any which are open!
